contributors/design-proposals/statefulset-update.md

```diff
@@ -0,0 +1,816 @@
+    // OnDeleteStatefulSetStrategyType triggers the legacy behavior. Version 
+    // tracking and ordered rolling restarts are disabled. Pods are recreated 
+    // from the StatefulSetSpec when they are manually deleted.
+    OnDeleteStatefulSetStrategyType = "OnDelete"
+)
+
+// PartitionedStatefulSet contains the parameters used with the 
+// PartitionedStatefulSetStrategyType.
+type PartitionedStatefulSet struct {
+    // Ordinal indicates the ordinal at which the StatefulSet should be 
+    // partitioned.
+    Ordianl uint32
+}
+
+type StatefulSetSpec struct {
+    // Replicas, Selector, Template, VolumeClaimsTemplate, and ServiceName 
+    // omitted for brevity.
+
+	// UpdateStrategy indicates the StatefulSetUpdateStrategy that will be 
+	// employed to update Pods in the StatefulSet when a revision is made to 
+	// Template or VolumeClaimsTemplate.
+	UpdateStrategy StatefulSetUpdateStrategy `json:"updateStrategy,omitempty`
+
+	// RevisionHistoryLimit is the maximum number of PodTemplates that will 
+	// be maintained in the StatefulSet's revision history. The revision history
+	// consists of all revisions not represented by a currently applied 
+	// StatefulSetSpec version. The default value is 2.
+	RevisionHisotryLimit int32 `json:revisionHistoryLimit,omitempty`
+}
+```
+
+The following modifications will be made to the StatefulSetStatus API object.
+
+```go
+ type StatefulSetStatus struct {
+    // ObservedGeneration and Replicas fields are omitted for brevity.
+
+ 	// CurrentVersion, if not empty, indicates the version of PodSpecTemplate, 
+ 	// VolumeClaimsTemplate tuple used to generate Pods in the sequence
+ 	// [0,CurrentReplicas).
+ 	CurrentVersion string `json:"currentVersion,omitempty"`
+ 	
+ 	// UpdatedVersion, if not empty, indicates the version of PodSpecTemplate, 
+    // VolumeClaimsTemplate tuple used to generate Pods in the sequence
+    // [Replicas-UpdatedReplicas,Replicas)
+ 	UpdateVersion string `json:"updateVersion,omitempty"`
+ 	
+ 	// ReadyReplicas is the current number of Pods, created by the StatefulSet
+    // controller, that have a Status of Running and a Ready Condition.
+ 	ReadyReplicas int32 `json:"readyReplicas,omitempty"`
+ 	
+ 	// CurrentReplicas is the number of Pods created by the StatefulSet 
+    // controller from the PodTemplateSpec, VolumeClaimsTemplate tuple indicated 
+    // by CurrentVersion.
+ 	CurrentReplicas int32 `json:"currentReplicas,omitempty"`
+ 	
+ 	// UpdatedReplicas is the number of Pods created by the StatefulSet
+    // controller from the PodTemplateSpec, VolumeClaimsTemplate tuple indicated 
+    // by CurrentVersion.
+ 	UpdatedReplicas int32 `json:"taretReplicas,omitempty"`
+}
+```
+
+Additionally we introduce the following constant.
+
+```go
+// StatefulSetVersionLabel is the label used by StatefulSet controller to track
+// which version of StatefulSet's StatefulSetSpec was used generate a Pod.
+const StatefulSetVersionLabel = "StatefulSetVersion"
```

Comment:
`statefulset.kubernetes.io/revision` similar to https://github.com/kubernetes/kubernetes/blob/20457ee68dac5527f6d9307fddcd2bd92389e5d4/pkg/controller/deployment/util/deployment_util.go#L51 ?